crates/edr_solidity/src/vm_tracer.rs

```diff
@@ -0,0 +1,270 @@
+//! Naive Rust port of the `VmTracer` from Hardhat.
+
+use std::{cell::RefCell, rc::Rc};
+
+use edr_eth::Bytes;
+use edr_evm::{
+    alloy_primitives::U160,
+    trace::{BeforeMessage, Step},
+    ExecutionResult,
+};
+
+use crate::message_trace::{
+    BaseEvmMessageTrace, BaseMessageTrace, CallMessageTrace, CreateMessageTrace, EvmStep, ExitCode,
+    MessageTrace, MessageTraceStep, PrecompileMessageTrace,
+};
+
+type MessageTraceRefCell = Rc<RefCell<MessageTrace>>;
+
+/// Naive Rust port of the `VmTracer` from Hardhat.
+pub struct VmTracer {
+    tracing_steps: Vec<Step>,
+    message_traces: Vec<MessageTraceRefCell>,
+    last_error: Option<&'static str>,
+    max_precompile_number: u64,
+}
+
+impl Default for VmTracer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl VmTracer {
+    /// Creates a new [`VmTracer`].
+    pub const fn new() -> Self {
+        // TODO: temporarily hardcoded to remove the need of using ethereumjs' common
```

Comment:
I've ported this verbatim from JS. I'm not familiar with the code and the context around it but it seems like "max precompile number" should probably be shared in the EDR itself? Or would you rather we supply this information from JS until everything is ported over?

Reply:
Oh, this one has some history.

When we were using EthereumJS, we would do something like `common.precompiles.length` (I don't remember the exact API) to get the max precompile number. This was then used during tracing to figure out whether a call was a precompile call.

After we moved to EDR, I tried to remove as much EJS dependencies as possible. Some usages of `Common`, like this one, were trickier, so I just hardcoded a max value here. This is of course wrong: if you use an older hardfork that doesn't have the `0xA` precompile, and you make a call to that address, Hardhat will tell you it was a precompile call, when it wasn't. I think this is a fine, temporary compromise, because this only affects user messages, not execution.

Once we finish porting this to the EDR side, this should be improved so that this comes from the EDR EVM somehow.

Also notice that there is another problem with this: after RIP-7212, the idea that there is a max precompile number, and that every address below it is a precompile, is wrong. So we should improve that too.

@Xanewok do you mind opening an issue about this, and linking it in the comment?

Reply:
Opened https://github.com/NomicFoundation/edr/issues/565; feel free to reword it or add more context :+1:

Reply:
Linked to the issue in the TODO note as well in https://github.com/NomicFoundation/edr/pull/531/commits/b455234127468948e6511124e486c0cb428340cb.